File: macos/Flash/Data/BackupDiff.swift
// Copyright (c) 2026 Fengyuan Jiang
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation

struct DifferenceSummary: Equatable {
    let added: Int
    let changed: Int
    let unchanged: Int
    let localOnly: Int
}

struct BackupDifference: Equatable {
    let logs: DifferenceSummary
    let emotions: DifferenceSummary
}

enum BackupDiff {
    static func analyze(
        localLogs: [LogItem],
        localEmotions: [EmotionRecord],
        incomingLogs: [LogItem],
        incomingEmotions: [EmotionRecord]
    ) -> BackupDifference {
        BackupDifference(
            logs: summarize(local: localLogs, incoming: incomingLogs, id: \.id),
            emotions: summarize(local: localEmotions, incoming: incomingEmotions, id: \.id)
        )
    }

    private static func summarize<T: Equatable>(
        local: [T], incoming: [T], id: KeyPath<T, String>
    ) -> DifferenceSummary {
        let localByID = Dictionary(local.map { ($0[keyPath: id], $0) }, uniquingKeysWith: { _, last in last })
        let incomingByID = Dictionary(incoming.map { ($0[keyPath: id], $0) }, uniquingKeysWith: { _, last in last })
        var added = 0
        var changed = 0
        var unchanged = 0
        for (key, value) in incomingByID {
            guard let current = localByID[key] else { added += 1; continue }
            if current == value { unchanged += 1 } else { changed += 1 }
        }
        return DifferenceSummary(
            added: added,
            changed: changed,
            unchanged: unchanged,
            localOnly: localByID.keys.filter { incomingByID[$0] == nil }.count
        )
    }
}

File: android/app/src/main/java/com/flash/app/data/BackupDiff.kt
// Copyright (c) 2026 Fengyuan Jiang
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.flash.app.data

import com.flash.app.data.model.EmotionRecord
import com.flash.app.data.model.LogItem

data class DifferenceSummary(val added: Int, val changed: Int, val unchanged: Int, val localOnly: Int)
data class BackupDifference(val logs: DifferenceSummary, val emotions: DifferenceSummary)

object BackupDiff {
    fun analyze(
        localLogs: List<LogItem>,
        localEmotions: List<EmotionRecord>,
        incomingLogs: List<LogItem>,
        incomingEmotions: List<EmotionRecord>,
    ): BackupDifference = BackupDifference(
        summarize(localLogs, incomingLogs, LogItem::id),
        summarize(localEmotions, incomingEmotions, EmotionRecord::id),
    )

    private fun <T> summarize(local: List<T>, incoming: List<T>, id: (T) -> String): DifferenceSummary {
        val localById = local.associateBy(id)
        val incomingById = incoming.associateBy(id)
        var added = 0
        var changed = 0
        var unchanged = 0
        incomingById.forEach { (key, value) ->
            val current = localById[key]
            when {
                current == null -> added++
                current == value -> unchanged++
                else -> changed++
            }
        }
        return DifferenceSummary(added, changed, unchanged, localById.keys.count { it !in incomingById })
    }
}
